Fix issue where a DICOM server would stop accepting incoming connections if MaxClientsAllowed was configured and one or more initial connections never closed

amoerie opened the pull request:
Fixes #1468
Fixes #1594

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- The DicomServer.RemoveUnusedServices method had a subtle bug: when it was waiting for one or more services to complete (using Task.WhenAny), it did not take into account additional services that were connected since. 
- That means, as long as the initial connection(s) that were captured at the moment Task.WhenAny was started did not complete, the number of connected clients would never decrease. This eventually lead to reaching the maximum, at which point no new clients could connect, even if the total number of actual connected clients was well below the maximum.
- I've solved this issue by reimplementing the "max clients allowed" feature using a SemaphoreSlim, which seemed like a perfect fit for the use case. SemaphoreSlim is a lock with a configurable amount of threads allowed to hold the lock. 
- By switching the _isDisconnected flag to a TaskCompletionSource in DicomService (this is okay, the flag was never reset, and it makes sense that a DicomService can only ever disconnect once, since it represents an association), I was able to completely remove the Obsolete AsyncManualResetEvent under the Network namespace.
- I've added a `Action<DicomServiceOptions> configure` parameter to DicomServerFactory. This is something I would like to discuss. By adding this parameter, you can now modify the options that are provided to the server on startup. You can still modify the options after the server has started, but specifically for MaxClientsAllowed this is very strange. What do you do if there are 50 clients connected and then you set MaxClientsAllowed to 10? In the current implementation in this pull request, modifying MaxClientsAllowed after server startup has no effect. If you want to set it, you must use the `configure` parameter of DicomServerFactory or you must use `services.Configure<DicomServiceOptions>(o => ...)`. Be aware that this last option applies to all DICOM servers though, you can see it as changing the defaults.

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into development

amoerie committed:
Add unit test that demonstrates the problem

amoerie committed:
Add 'configure' parameter to DicomServerFactory

This is important because it allows us to modify the DicomServiceOptions before the values are 'captured' in the DicomServer

The setting 'MaxClientsAllowed' will no longer have any impact if it is changed after the DICOM server has started.

amoerie committed:
Fix DicomServerFactory not passing the 'configure' option properly

amoerie committed:
Use separate log prefixes for the 3 connections in the unit test

amoerie committed:
Reimplement DICOM server unused services cleanup

The new implementation uses a semaphore to enforce the max clients allowed setting, and a channel to send notifications from one thread to another that services are connecting

amoerie committed:
Remove last usage of obsolete AsyncManualResetEvent

amoerie commented:
@gofal @mrbean-bremen I think this issue just increased in urgency. Even if you don't use the MaxClientsAllowed setting, this bug can cause memory issues in the long run when

- one of the client connections never closes
- this happens to be the only connection when RemoveUnusedServices starts waiting

When this happens, for the remainder of the lifetime of the DicomServer, no DicomService instances are ever removed and disposed from the running services (`DicomServer._services`), which is a memory leak.

Surprisingly, this bug has been present for ages. It was introduced 6 years ago in this commit: https://github.com/fo-dicom/fo-dicom/commit/c2ee947457f1669f7242aae1d88cd95ce268c092#diff-875fb37c7d8a406fcb8ebf8fd2519e6b1bc266ecd8e2d783929daf38dced6e88 

Anyway, this bug manifests in different ways. If you use MaxClientsAllowed, your DicomServer simply stops working. If you don't use MaxClientsAllowed, your memory usage will keep increasing indefinitely. (We saw 140 000 instances of DicomService in one production environment)

Of course, you still need a "broken" client that never disconnects, but this is not so far fetched. I think Fellow Oak DICOM should handle this elegantly.

amoerie committed:
Handle DICOM server disposal in RemoveUnusedServices

amoerie committed:
Make RemoveUnusedServices more robust and add tests

amoerie committed:
Update changelog

amoerie committed:
Remove unit test that is too dependent on machine perf characteristics

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into max-clients-allowed

gofal committed:
update documentation and add breaking change to changelog

gofal commented:
The Action<DicomServiceOptions> configure parameter to DicomServerFactory as addition fits perfect into the common fluent api. 
Changing the maxclientsallowed after the application has started, would be solvable. eg. if there are already 50 connected and you change the maxclientsallowed to 10 then no furhter client is allowed until the number of connected clients lowered to 9. This would be very straight forward. But the current implementation will not support this.

What is important, is, to make it very clear fo the user, that certain options have to be set within the conctructor. I added some text to the property description.

To make it more obvious to the user, we could split up the Options into "Options" that are sill changeable while the server is running and "StartupOption" that can only be set in this action parameter of the factory method.
We could also consider to make StartupOptions inheriting from Options, then in the startupfactory you can change all the options not only the specific unmutual ones. and the user accesses this private filed through the property of type "DicomServiceOptions" and therefore can only change those that are changeable while the server is running.

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into max-clients-allowed

amoerie committed:
Merge server startup fix

Slightly optimize the fix too: instead of creating a task in every RemoveUnusedServiceAsync loop iteration, only create the task once in the constructor

Secondly, instead of attaching a cancellation token listener, just set the task to completed in the one single place where the cancellation token is canceled.

amoerie committed:
Add DicomServerOptions

Move TcpNoDelay and MaxClientsAllowed to DicomServerOptions
Add TcpNoDelay to DicomClientOptions

amoerie committed:
Update changelog

amoerie committed:
Update dependency injection to include DicomServerOptions

amoerie committed:
Update changelog

amoerie commented:
I liked your idea of making it more explicit which options are "set once" and which options are still changeable after the server is running. I started out implementing your idea of having DicomServerStartupOptions  that inherit from DicomServiceOptions, but I had some concerns:

- Cloning in combination with `IOptions<DicomServiceOptions>` was not very straightforward
- TcpNoDelay and MaxClientsAllowed are actually not related to DicomService at all.

I tried going in a slightly different direction, and the result came out quite nice imho:

- I added DicomServerOptions (DicomClientOptions already exists, so I like the parallel between these)
- I moved MaxClientsAllowed to DicomServerOptions
- I moved TcpNoDelay to DicomServerOptions and DicomClientOptions
- The new 'configure' parameter of DicomServerFactory is now of type `Action<DicomServerOptions>`

DicomServerOptions is not exposed on DicomServer, since modifying the options after the server has started does nothing.

Edit: since the TCP buffer sizes PR has been merged, I did the same for TcpReceiveBufferSize and TcpSendBufferSize. Please let me know what you think. :-)

amoerie committed:
Fix issue caused by inlining variable

gofal committed:
Merge branch 'development' into max-clients-allowed

amoerie committed:
Move TcpReceiveBufferSize and TcpSendBufferSize to DicomClientOptions and DicomServerOptions

gofal commented on FO-DICOM.Core/Network/DicomServer.cs:
```diff
@@ -273,33 +278,37 @@ private async Task ListenForConnectionsAsync()
 
+                var maxClientsAllowed = _serverOptions.MaxClientsAllowed;
+                var noDelay = _serverOptions.TcpNoDelay;
+                var receiveBufferSize = _serverOptions.TcpReceiveBufferSize;
```
why is it necessary to store the ReceiveBufferSize in this local variable on startup?
if later within the loop you call 
`var networkStream = await listener
                        .AcceptNetworkStreamAsync(_tlsAcceptor, noDelay,Options.TcpReceiveBufferSize, Options.TcpSendBufferSize, Logger, _cancellationToken)`
then you can change the setting while the server is running and will be applied on the next connection.

amoerie replied:
I thought about this and felt like it was not really necessary to support this, but I didn't feel strongly about it, and you are right that it is trivial to make this scenario work. I
've moved the TCP settings back to DicomServiceOptions and inlined the variables so they are applied in real time for new connections. 👍

gofal commented on FO-DICOM.Core/Network/DicomServerOptions.cs:
```diff
@@ -0,0 +1,44 @@
+        /// If not configured, the default value of 8192 bytes will be used
+        /// </summary>
+        /// <seealso cref="System.Net.Sockets.TcpClient.ReceiveBufferSize"/>
+        public int? TcpReceiveBufferSize { get; set; }
```
The property `TcpReceiveBuffersize` currently is in DicomServerOptions as well as in DicomServiceOptions. In your comment of this PR you told, you had moved the MaxClientsAllowed and the TcpNoDelay into the new DicomServerOptions class. Does that mean, that these two buffersize-options should be removed here?

amoerie replied:
Ah yes they should not be in the service options. Thanks!

gofal replied:
I meant, they should not be in the server options. Because it would be fine changing the values even after the server has startet. the Server options are only accessable as action in the factory method and was intended for options that need to be set before the server is starting.

amoerie replied:
Yes taking your other comment into account, this makes more sense. I've made the requested changes. 👍

amoerie committed:
Merge branch 'development' of github.com:fo-dicom/fo-dicom into max-clients-allowed

amoerie committed:
Move TCP options back to DicomServiceOptions

These options can be applied to both a client and a server, and they can be modified while a server or client is running. Therefore, they can stay in DicomServiceOptions.

amoerie committed:
Use compile time constant log messages

amoerie committed:
Remove breaking changes from changelog

amoerie committed:
Merge branch 'development' into max-clients-allowed
